snapshots: Fix snapshot-ctl to include snapshots of dropped tables

StarostaGit opened the pull request:
Snapshot-ctl methods fetch information about snapshots from column family objects. The problem with this is that we get rid
of these objects once the table gets dropped, while the snapshots might still be present (the auto_snapshot option is specifically made to create this kind of situation). This commit switches from relying on column family interface to scanning every datadir that the database knows of in search for "snapshots" folders.

This PR is rebased onto bhalevy@2804fc4 in order to use `sstables::snapshots_dir` introduced by @bhalevy 

Fixes #3463
Closes #7122

bhalevy commented on database.cc:
```diff
@@ -2268,6 +2268,69 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
     });
 }
 
+future<std::unordered_map<sstring, std::unordered_map<sstring, db::snapshot_ctl::snapshot_details>>> database::get_snapshot_details() {
```
please use a coroutine

bhalevy reviewed (commented):
@nyh, please review also from the functionality perspective since you commented on this issue on the mailing list.

Talking to @avikivity, he wants to keep the CQL syntax thin with this respect and potentially use a (virtual) table to represent detached snapshots.

haaawk commented:
@xemul @nyh  ping

haaawk commented:
@avikivity is the scan of all the dirs ok for you or do we need to cache the info about all the snapshot somewhere? It seems that this operation is rare and it's better to scan on those rare occasions the user runs the command. What do you think?

haaawk commented:
> Talking to @avikivity, he wants to keep the CQL syntax thin with this respect and potentially use a (virtual) table to represent detached snapshots.

What do you mean @bhalevy? This is a nodetool command and CQL syntax is not changed at all, is it?

nyh commented:
> Snapshot-ctl methods fetch information about snapshots from column family objects. The problem with this is that we get rid of these objects once the table gets dropped, while the snapshots might still be present (the auto_snapshot option is specifically made to create this kind of situation). 

Yes, it is ironic that the `auto_snapshot` option was meant to allow undoing deletion, yet we don't offer any mechanism to actually undo the deletion - or any guarantee of it will actually be possible (e.g., in case of cluster changes). In https://github.com/scylladb/scylla/issues/7123 I proposed what I believe to be a better way, but I agree that having a better way doesn't mean we can't fix a bit the code that we have now.

> Fixes [scylladb/scylla-enterprise#1884](https://github.com/scylladb/scylla-enterprise/issues/1884)

This specific link isn't visible to the public. Maybe you meant fixes https://github.com/scylladb/scylla/issues/3463 ?

nyh commented:
Please note that @bhalevy also has issue https://github.com/scylladb/scylla/issues/7122 suggesting a **different** approach - an alternative approach which you mentioned in your DISCLAIMER. If we decide to go with your approach, please add a "Closes #7122" to your commit message so that Benny's suggestion will be closed.

nyh commented on database.cc:
```diff
@@ -2268,6 +2268,69 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
+
+                return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
```
Do we have to open-code this filename parsing here - don't we already have it in the sstable code and can call that?
I'm worried that the directory structure or filename structure might change between sstable versions. In fact, didn't it change since KA? Or maybe we don't care about supporting KA any more?

StarostaGit replied:
I wasn't aware of any function that would already provide cf_name <-> cf_dir_name translation, but I'll look some more.

bhalevy replied:
We have a mapping from a sstable file name to the sstable attributes (version, generation, etc.) in the sstable class but I don't remember having something like that for tables.

There's `keyspace::column_family_directory` that maps `<cf_name, uuid>` into the directory entry name
but I saw no reverse mapping of that.
Maybe the mapping should be implemented, along side ``keyspace::column_family_directory` so it would be easier to change them together if and when this happens.

nyh commented on database.cc:
```diff
@@ -2268,6 +2268,69 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
+
+                        return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
+                            sstring snapshot_name = de.name;
+                            sstring cf_map_key = ks_name + ":" + cf_name;
```
Can we have a pair of strings instead of combining them with a colon?
While the colon character is technically not allowed in table names in CQL, it's not outlawed by our underlying code, and Alternator does uses the colon character in some of the table names (for materialized views).

StarostaGit replied:
yeah, sure

nyh reviewed (changes requested):
I left a couple of small requests, but I also have a question about the general approach:
If I understand your code correctly, even when the table is alive, you no longer call its `get_snapshot_details()`.  So can we remove this function?

Moreover, I'm asking myself (and now asking you) - from where did that old function get its information? Did it also scan the local disk to look for these directories? If this is the case, why did you need to write new code to do this instead of re-using the old code that did it?

avikivity commented:
> @avikivity is the scan of all the dirs ok for you or do we need to cache the info about all the snapshot somewhere? It seems that this operation is rare and it's better to scan on those rare occasions the user runs the command. What do you think?

Agree that real-time scan is best. The user may delete snapshots at any time, so any copy of the information can be invalid.

StarostaGit commented:
@nyh 
> If I understand your code correctly, even when the table is alive, you no longer call its get_snapshot_details(). So can we remove this function?
It still gets used elsewhere, e.g. in the  `column_family` section of our API
> Moreover, I'm asking myself (and now asking you) - from where did that old function get its information? Did it also scan the local disk to look for these directories? If this is the case, why did you need to write new code to do this instead of re-using the old code that did it?
It does a very similar thing, my code is based on the cf specific version, main difference being that the `table` objects know some info about their data dirs and whatnot. I couldn't just call the old functions precisely because they were cf specific - and we want to include snapshots of tables we no longer keep info about.

All that being said, I could add table and keyspace filters to the new db method if we agree there's no point in keeping both the table specific and the general snapshot getting methods in our code. There is a theoretical advantage to the `table::get_snapshot_details()` method over the generic one I added - it's more dir structure and naming agnostic. The `table` object knows its datadir (regardless of any naming conventions and any changes to them) and it guarantees that what we're looking through is a directory of one of scylla's column families. Whereas in the general approach that I added I have to rely on a directory naming convention to extract column family names from dir names and, theoretically, someone could manually add a dir with a snapshots folder that we will now scan through.

bhalevy commented on database.cc:
```diff
@@ -2268,6 +2268,70 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
+                    fs::path cf_dir = parent_dir / de.name;
+
+                    return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
```
what about other subdirectories like "staging", "upload", etc?
please see my `compaction: quarantine invalid sstables` series on scylladb-dev the mailing list
as I'm adding yet another subdirectory.
and in particular, `[PATCH v4 01/11] sstables: define symbolic names for table subdirectories`
that introduces `sstables::snapshots_dir` that you could use for listing only the snapshots subdirectory.

StarostaGit replied:
Snapshots can only appear in the snapshots directory, can't they? So any other directories are not relevant to us.

As for the `sstables::snapshots_dir` it does look like a nice thing to use. The patch did not go in yet, did it? Can I use it before it's introduced (e.g. add the `constexpr ... snapshots_dir = ...` myself)?

bhalevy replied:
no, it didn't go in yet
Feel free to cherry pick the needed patch from my branch:

https://github.com/bhalevy/scylla/commit/2804fc4ee07b7eb9f20eea3a10c209b360a4bb92

bhalevy replied:
The whole branch is https://github.com/bhalevy/scylla/commits/quarantine-invalid-sstables

StarostaGit commented on database.cc:
```diff
@@ -2268,6 +2268,70 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
 }
 
+future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> database::get_snapshot_details() {
+    co_return co_await seastar::async([this] () {
```
We need to scan through the directories (calling `lister::scan_dir`) on disk, which is currently operating on futures. I'll add a comment above the async

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +199,13 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
-                        for (auto&& snap_map: map) {
-                            local_total += snap_map.second.live;
+                return db.get_snapshot_details().then([&local_total] (auto&& map) {
+                    for (const auto& [snapshot, cf_map] : map) {
```
Would it make sense to substitute `snapshot` with `std::ignore`?

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +199,13 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
-                            local_total += snap_map.second.live;
+                return db.get_snapshot_details().then([&local_total] (auto&& map) {
+                    for (const auto& [snapshot, cf_map] : map) {
+                        for (const auto& [cf_id, details] : cf_map) {
```
Would it make sense to substitute `cf_id` with `std::ignore`?

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +199,13 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
-                    return cf_pair.second->get_snapshot_details().then([&local_total] (auto map) {
-                        for (auto&& snap_map: map) {
-                            local_total += snap_map.second.live;
+                return db.get_snapshot_details().then([&local_total] (auto&& map) {
```
Is it possible for the map to be very big and cause the following nested loops to create a reactor stall? If so, maybe it would make sense to use one of seastar loops and check need_preempt() inside?

StarostaGit replied:
I don't think it can, the size of the map is equal to the number of snapshots - which shouldn't be that big?

haaawk replied:
Those are snapshots for all tables, no? Meaning that there can be thousands of them.

StarostaGit replied:
Well yeah, but is thousands a very big number tho?

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -144,7 +144,7 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
 future<std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>>>
 snapshot_ctl::get_snapshot_details() {
-    using cf_snapshot_map = std::unordered_map<utils::UUID, column_family::snapshot_details>;
+    using cf_snapshot_map = std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash>;
```
A comment describing what this std::pair<string, string> represents would make it easier to understand the code

haaawk commented on database.hh:
```diff
@@ -1568,6 +1569,8 @@ public:
      */
     future<> clear_snapshot(sstring tag, std::vector<sstring> keyspace_names, const sstring& table_name);
 
+    future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> get_snapshot_details();
```
A comment describing the returned type would make it easier to understand the code.

StarostaGit replied:
I was considering introducing an alias here with `using`, but wasn't quite sure whether we want to have this exported to other translation units (since it's a header file)

haaawk replied:
I think a comment would be good enough

kbr- replied:
So we don't get neither comments nor `using`? :(

haaawk commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+    // We require a seastar thread to run `lister::scan_dir`s
+    co_return co_await seastar::async([this] () {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        auto dirs_only_entries_ptr =
```
Would a std::unique_ptr be enough here instead of lw_shared_ptr?

haaawk commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+    // We require a seastar thread to run `lister::scan_dir`s
+    co_return co_await seastar::async([this] () {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        auto dirs_only_entries_ptr =
```
Why does it have to be allocated on a heap? Can't it be stack allocated?

haaawk replied:
Like details

StarostaGit replied:
Hm, I based this part on other places in our codebase where we search through the directories and they all use a pointer. But tbh now that you mention it, I have no idea why. Changed it to automatic lifetime and it seems to be working fine

haaawk commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                                        size = 0;
+                                    }
+                                    return io_check(file_size, (fs::path(cf_dir) / name).native()).then_wrapped([cf_map_key, &details, snapshot_name, size] (auto fut) {
+                                        try {
+                                            // File exists in the main SSTable directory. Snapshots are not contributing to size
+                                            fut.get0();
+                                        } catch (std::system_error& e) {
+                                            if (e.code() != std::error_code(ENOENT, std::system_category())) {
+                                                throw;
+                                            }
+                                            details.at(snapshot_name).at(cf_map_key).live += size;
+                                        }
```
Is there a way to check that a file is not present in the sstable directory without throwing exceptions that are slow?

haaawk replied:
On a second thought this is probably alright and we can do improvements as a follow up if needed.

haaawk commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                                    // All the others should just generate an exception: there is something wrong, so don't blindly
+                                    // add it to the size.
+                                    if (name != "manifest.json" && name != "schema.cql") {
+                                        sstables::entry_descriptor::make_descriptor(snapshot_dir.native(), name);
```
Why is this called when its returned type is ignored?

StarostaGit replied:
looks like a leftover from a previous version, removed it

haaawk replied:
I just want to make sure that it's not validating something and throwing exception when something's wrong. I can see that it was called like this in the past as well https://github.com/scylladb/scylla/blob/3226c5bf9d6db6c3f3c57008aacab9cab8c57fc9/table.cc#L1473 so I'm not sure if it's an old code that should be removed or something important. @bhalevy do you know?

haaawk replied:
All other usages of the function do something with the return type.

haaawk commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+    // We require a seastar thread to run `lister::scan_dir`s
+    co_return co_await seastar::async([this] () {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        // auto dirs_only_entries_ptr =
```
???

kbr- commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                    return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
+
+                        return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
```
Why suddenty not performing `move`s on `ks_name` etc. in the lambda capture?

kbr- replied:
BTW. The logic inside this `scan_dir` (which handles a single snapshot) could really benefit (in terms of readability) from being extracted to a separate function.

StarostaGit replied:
> Why suddenty not performing moves on ks_name etc. in the lambda capture?

Nested `scan_dirs` are like nested loops, so we don't want to move a value from an outer loop into an inner loop, cause we will still need it in the next iteration of the outer one. However, I might be able to pass by reference there and avoid excessive copies anyway.

kbr- commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                            sstring snapshot_name = de.name;
+                            std::pair<sstring, sstring> cf_map_key = { ks_name, cf_name };
+                            std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash> cf_map = {{ cf_map_key, {0, 0, cf_name, ks_name} }};
+                            details.emplace(snapshot_name, std::move(cf_map));
```
Why not do this at the end, after populating `cf_map`?
You wouldn't have to call `details.at(snapshot_name)` below.

StarostaGit replied:
Since we only write into one key, I switched from passing around the `details` map and the keys to just passing around an iterator to the `cf_map`'s element. This also avoids unnecessary lookups

kbr- commented on database.cc:
```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                                    }
+                                    return io_check(file_size, (fs::path(cf_dir) / name).native()).then_wrapped([cf_map_key, &details, snapshot_name, size] (auto fut) {
+                                        try {
+                                            // File exists in the main SSTable directory. Snapshots are not contributing to size
```
"File exists in the main SSTable directory"
If I understand correctly, we don't know that yet, we will know it after we leave the `catch` block below. The comment at this place is a little confusing, IMO it should be moved after the `catch`.

Also "Snapshots are not contributing to size" - can you elaborate why? (by continuing this comment)

kbr- reviewed (changes requested):
This really needs a test which creates a bunch of tables, deletes some of them, snapshots the others, and check if every expected snapshot is present. Preferably the test should be part of the test.py suite (using cql-pytest). If this is problematic, a dtest.

kbr- commented on database.cc:
```diff
@@ -2321,6 +2321,64 @@ future<> database::drain() {
+            // KS directory
+            sstring ks_name = de.name;
+
+            co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable -> future<> {
```
When you're using `co_await`, everything in the coroutine's frame is kept alive. This distinguishes `co_return co_await e` from `return e` (where `e` is an expression giving a future): when you `return`, the function's frame is destroyed, but when you `co_return co_await`, you first wait for `e` to finish (i.e. the future must resolve), and only then the frame is destroyed. So for `co_return co_await`, you can depend on local variables to remain valid, unlike when you `return`.

On the other hand `co_await` turns your function into a coroutine, causing an additional heap allocation (coroutine frame).

In this case you would probably be fine with a simple `return` since you don't need to keep anything alive.
However this is not performance critical code so I guess `co_return co_await` is also OK, and it makes the code more consistent. So depends on your preference.

kbr- replied:
Same comment applies to the `co_return co_await` expressions below - you don't capture references to local variables of your coroutines, so they could as well be functions.

StarostaGit replied:
I left it as is for consistency

StarostaGit commented:
Left to address:
* new feedback about coroutines
* test
* explain size calculations - I have no idea about the reasoning behind it, it was taken from the `get_snapshot_details()` method for tables. I have look into it a bit if we want explanations

StarostaGit commented:
Removed unnecessary `co_retur;` and added a unit test

StarostaGit commented:
Transformed nested loops in `true_snapshot_size()` to seastar's `do_until`s with `need_preempt()`

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +201,26 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
+                    auto cf_map_it = map.cbegin();
+                    auto cf_map_end = map.cend();
+                    return do_until([&cf_map_it, &cf_map_end] { return cf_map_it == cf_map_end; }, [&cf_map_it, &cf_map_end, &local_total] {
+                        if (cf_map_it == cf_map_end || need_preempt()) {
```
Won't `map` be destroyed by this point? What's keeping it alive?

haaawk replied:
Same with cf_map_it and cf_map_end. Aren't they automatic variables that are destroyed on return?

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +201,26 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
+                        auto details_it = cf_map_it->second.cbegin();
+                        auto details_end = cf_map_it->second.cend();
+                        return do_until([&details_it, &details_end] { return details_it == details_end; }, [&details_it, &details_end, &local_total] {
+                            if (details_it == details_end || need_preempt()) {
```
Same question here. What's keeping details_it and details_end from being destroyed?

kbr- commented on db/snapshot-ctl.cc:
```diff
@@ -209,13 +201,26 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
+                return db.get_snapshot_details().then([&local_total] (auto&& map) {
+                    auto cf_map_it = map.cbegin();
+                    auto cf_map_end = map.cend();
+                    return do_until([&cf_map_it, &cf_map_end] { return cf_map_it == cf_map_end; }, [&cf_map_it, &cf_map_end, &local_total] {
```
The seastar loops have preemption checking built-in, why the manual `need_preempt()`s?

Also why not use a normal loop inside a coroutine combined with `co_await coroutine::maybe_yield()`?

StarostaGit commented:
turned both `true_snapshot_size()` and `get_snapshot_details()` into coroutines

kbr- commented on db/snapshot-ctl.cc:
```diff
@@ -174,51 +181,41 @@ snapshot_ctl::get_snapshot_details() {
+                        local_total += details.live;
+
+                        if (need_preempt()) {
+                            co_await coroutine::maybe_yield();
```
`coroutine::maybe_yield()` checks if preemption is needed. That's the whole point. You don't need the `if` :)

StarostaGit replied:
huh, I searched through our codebase for `maybe_yield()` and that `if` - `yield` pair was the first result xd
but good to know

kbr- commented on db/snapshot-ctl.cc:
```diff
@@ -174,51 +181,41 @@ snapshot_ctl::get_snapshot_details() {
-                            local_total += snap_map.second.live;
+    co_return co_await run_snapshot_list_operation([this] () -> future<int64_t> {
+        co_return co_await _db.map_reduce(adder<int64_t>(), [] (database& db) -> future<int64_t> {
+            co_return co_await do_with(int64_t(0), [&db] (auto& local_total) -> future<int64_t> {
```
Since you turned this into a coroutine, you don't need the `do_with` anymore.

StarostaGit replied:
ah, right

StarostaGit commented:
removed do_with in a coroutine

kbr- commented on database.cc:
```diff
@@ -2321,6 +2321,69 @@ future<> database::drain() {
+    return directory_name.substr(0, directory_name.find("-"));
+}
+
+future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> database::get_snapshot_details() {
```
As we talked on slack: this could return a simple vector of structs, you never use the `unordered_map` structure.

kbr- reviewed (changes requested):
BTW. when you push a change, please write a comment with a short summary of the changes.

EDIT: I see you usually do that, so nvm
(but still I don't see a comment for the last push; please try to remember about it - it's useful for reviewers to know that something has changed)

StarostaGit commented:
Removed the unnecessary `mutable`s and changed the return type of `get_snapshot_details()` to a vector.
The `snapshot_reducer` now uses an `unordered_set` to eliminate the copies of snapshot details that come from map

kbr- commented on db/snapshot-ctl.cc:
```diff
@@ -144,82 +147,57 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
+        return utils::tuple_hash()(std::make_tuple(sdr.snapshot_name, sdr.details.ks, sdr.details.cf));
+    };
+
+    // only one entry per {snapshot_name, ks, cf} triplet
```
Could it happen that you get multiple entries with same snapshot name, ks and cf, but different `total`/`live`, because you're using default definitions of `operator==`? This doesn't directly follow from the definition of the map (because of `operator==`), but I guess it's simply impossible to have two such snapshots.

StarostaGit replied:
Yes, this is my understanding of the original implementation. There should be no snapshots that differ only in `total`/`live` as they are merely copies of each other that we just have to reduce into one. That's what the original was doing, the only difference being that it retained the last one, while I keep the first one.

kbr- commented on test/boost/database_test.cc:
```diff
@@ -365,39 +365,48 @@ SEASTAR_THREAD_TEST_CASE(test_distributed_loader_with_pending_delete) {
-            e.execute_cql("insert into cf (p1, c1, c2, r1) values ('key1', 3, 2, 3);").get();
+        do_with_cql_env_thread([cf_names = std::move(cf_names), func = std::move(func)] (cql_test_env& e) {
+            for (auto&& cf_name : cf_names) {
+                e.create_table([cf_name = std::move(cf_name)] (std::string_view ks_name) {
```
You got lucky here though: the lambda we're currently in is not `mutable`, so `cf_name` is actually a `const` reference...
Please drop the `move` here, it doesn't do anything right now but it's confusing.

kbr- commented on test/boost/database_test.cc:
```diff
@@ -476,8 +485,34 @@ SEASTAR_TEST_CASE(snapshot_list_okay) {
+
+        const auto& sd = details.front().details;
+        BOOST_REQUIRE_GT(sd.live, 0);
+        BOOST_REQUIRE_EQUAL(sd.total, sd.live);
```
Maybe we could make `total` non-zero by inserting some data?

StarostaGit replied:
Notice that I use `BOOST_REQUIRE_GT`. `sd.live` should be non-zero (greater than zero in fact) and then `sd.total` should match that

StarostaGit commented on test/boost/database_test.cc:
```diff
@@ -476,8 +485,34 @@ SEASTAR_TEST_CASE(snapshot_list_okay) {
+        details = e.local_db().get_snapshot_details().get0();
+        BOOST_REQUIRE_EQUAL(details.size(), 4);
+
+        return make_ready_future<>();
```
Sure, no problem. I omitted it because it's checked in a different test, but there's no harm checking it here as well.

StarostaGit commented:
Addressed feedback & rebased

haaawk commented on db/snapshot-ctl.cc:
```diff
@@ -144,82 +147,57 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
+
+        snapshot_map result;
+        for (auto&& r : collected) {
+            result[r.snapshot_name].emplace_back(std::move(r.details));
```
Does it make sense to add `co_await coroutine::maybe_yield();` here as well?

StarostaGit replied:
Yeah, it does iterate over the whole set of snapshots

haaawk commented on db/snapshot-ctl.hh:
```diff
@@ -61,6 +61,8 @@ public:
         sstring cf;
         sstring ks;
+
+        bool operator==(const snapshot_details& other) const = default;
```
ultra NIT: `other` can be removed as the variable is never used.

haaawk commented on database.hh:
```diff
@@ -1570,7 +1571,15 @@ public:
+    struct snapshot_details_result {
+        sstring snapshot_name;
+        db::snapshot_ctl::snapshot_details details;
+        bool operator==(const snapshot_details_result& other) const = default;
```
ultra NIT: you don't need to name the unused argument, do you?

haaawk reviewed (commented):
LGTM only one question about missing co_await coroutine::maybe_yield();

haaawk reviewed (approved):
LGTM @kbr- what do you think? Can we finalise this PR?

kbr- commented on database.cc:
```diff
@@ -2326,6 +2326,68 @@ future<> database::drain() {
+                            // All the others should just generate an exception: there is something wrong, so don't blindly
+                            // add it to the size.
+                            if (de.name != "manifest.json" && de.name != "schema.cql") {
+                                details.back().details.total += size;
```
I only hope `scan_dir` doesn't scan the files in parallel, otherwise `details.back()` may no longer refer to what we want.

kbr- replied:
>  I've switched to index-based access to be safe for sure

You're still using `details.back()` in one place.

BTW. Why not do the `push_back` after the `scan_dir`?

StarostaGit replied:
switched to doing just that

StarostaGit commented:
Switched from `back()` vector element access to index-based to avoid parallelism issues

StarostaGit commented:
comment addressed

bhalevy commented on database.cc:
```diff
@@ -2326,6 +2326,70 @@ future<> database::drain() {
+    std::vector<database::snapshot_details_result> details;
+
+    for (auto& datadir : data_dirs) {
+        co_await lister::scan_dir(datadir, dirs_only_entries, [this, &dirs_only_entries, &details] (fs::path parent_dir, directory_entry de) -> future<> {
```
https://github.com/scylladb/scylla/pull/9835 can simplify this nested scan a lot

haaawk replied:
@StarostaGit has left Scylla so I think we will probably just merge this PR. Or maybe you would like to pick it up and rebase on top on #9835 @bhalevy ?

bhalevy replied:
@haaawk #9835 can be done as a followup. it doesn't fix anything. We can proceed with this PR as is as far as I'm concerned.

@nyh any objections?

nyh replied:
@bhalevy @haaawk unfortunately there are new conflicts. Can one of you please rebase and fix these conflicts? I'll commit the results.

bhalevy replied:
@nyh, I'm rebasing and retesting ti.
Will submit a new PR since I can't (and don't want to) force-push into @StarostaGit's repo.

bhalevy commented:
Closing in favor of https://github.com/scylladb/scylla/pull/9884
